### PR TITLE
FEATURE: Make multiple AuthenticationProviders possible

### PR DESCRIPTION
**Pull request by gerdemann**
With this change it is possible to use multiple AuthenticationProviders
for the Neos backend.

### CODE REVIEW

File: TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php [hunk 1]
```diff
@@ -168,7 +168,7 @@ public function getUser($username, $authenticationProviderName = null)
         }
         $user = $this->partyService->getAssignedPartyOfAccount($account);
         if (!$user instanceof User) {
-            throw new Exception(sprintf('Unexpected user type "%s". An account with the identifier "%s" exists, but the corresponding party is not a Neos User.', get_class($user), $username), 1422270948);
+            return null;
```
Comment (bwaidelich) [inline TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php]: Why this change? Doesn't it break the behavior of the public API?

File: TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php [hunk 2]
```diff
@@ -214,7 +214,25 @@ public function getCurrentUser()
...
+        return null;
+    }
+
+    /**
+     * Returns the username of the currently logged in user, if any
+     *
+     * @return string The username or null
+     * @api
+     */
+    public function getCurrentUsername()
```
Comment (bwaidelich) [inline TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php]: Is this related?

File: TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php [hunk 3]
```diff
@@ -214,7 +214,25 @@ public function getCurrentUser()
         if ($this->securityContext->canBeInitialized() === true) {
             $account = $this->securityContext->getAccount();
```
Comment (bwaidelich) [inline TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php]: This is not introduced with this change, but IMO _that_ is actually the limiting factor here:
`Security\Context::getAccount()` just returns the first authenticated account, whichever the first account is and whatever authentication provider it belongs to.
I think we absolutely have to make this explicit and use `Security\Context::getAccountByAuthenticationProviderName()` instead

**Comment by bwaidelich** [comment]
I didn't test it, but I'm pretty sure this will break existing installations that use more than one Auth providers/user models because it would just try to map the first authenticated account to some user - whether that is a Neos user or not

**Comment by kitsunet** [comment]
What is the state here? @gerdemann could you answer the questions, do you plan to follow up on this change?

File: TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php (hunk 2 above)
Comment (gerdemann) [inline TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php]: Yes, that is needed to get the current username on the basis of the SecurityContext.

File: TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php (hunk 1 above)
Comment (gerdemann) [inline TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php]: I don't know if the public API breaks, but this is necessary so that the rest of the method will be executed.

**Comment by gerdemann** [comment]
Unfortunately, I have no time at the moment to take care of this pull request.
Nevertheless, it would be great if Neos could have several AuthenticationProviders for the backend.

**Comment by radmiraal** [comment]
Great to see work in this area, just one remark: it is still incomplete, on user generation a unique workspace identifier should be stored as well (to make it BC compatible the currently used scheme could be stored on the users preferences by a migration).

This is because the backend redirection service and retrieval of the workspacename would conflict if the same account identifier would be used over different providers (which is a possibility and valid usecase).

**Comment by kitsunet** [comment]
@radmiraal is this in conflict with the work you recently started?

**Comment by radmiraal** [comment]
@kitsunet nope, not at first sight. It might actually be the fix for one of the comments on untouched code that came up by reviewing my PR... ;) I'll pull this change into mine and take care of them together if @gerdemann doesn't mind

**Comment by kitsunet** [comment]
Can @bwaidelich and maybe @robertlemke have a short look? I would like to get this merged as it makes a lot of sense.

**Comment by gerdemann** [comment]
@kitsunet @bwaidelich @robertlemke If you give me hints and push me in the right direction, I can help you.

**Comment by kitsunet** [comment]
Ping again for @bwaidelich and @robertlemke

File: TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php (hunk 1 above)
Comment (bwaidelich) [reply TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php]: OK, I think we need to differentiate here along the lines of
```php
if ($user === null) {
    return null;
}
if (!$user instanceof User) {
    throw new Exception(sprintf('Unexpected user type "%s". An account with the identifier "%s" exists, but the corresponding party is not a Neos User.', get_class($user), $username), 1422270948);
}
```

File: TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php (hunk 2 above)
Comment (bwaidelich) [reply TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/UserService.php]: I'm still skeptical about this one: This is the UserService for Neos Users, but that method might return an unrelated Account identifier

**Comment by bwaidelich** [comment]
We really should fix support for multiple auth providers, but I'm a little skeptical about the approach..
Can we have a discussion about this and create a ticket with the issues we want to solve?

Because IMO we need a more "integral" solution for this. Having `UserService::getCurrentUser()` return `null` if a different auth provider was used, will probably break a lot of parts (e.g. `WorkspaceService::getAllowedTargetWorkspaces()`, `LocaleService`, ...).

Instead I think we need to extend the `\Neos\ContentRepository\Domain\Model\UserInterface` and refer to that everywhere and then extend the UserService to return an instance even for other auth providers (that support it).

Maybe we can get rid of one of the two UserServices at the same time
